pallene/typedecl.lua

```diff
@@ -82,4 +82,31 @@ function typedecl.declare(module, mod_name, type_name, constructors)
     end
 end
 
+-- Check if the 2nd tag is a prefix of the first.
+-- If appending a "." to the second argument makes it a
+-- prefix of the first, captures the rest of the string and
+-- returns it
+--
+-- for example:
+-- ```
+-- typedecl.match_tag("ast.Exp.Bool", "ast.Exp")
+-- ```
+-- would return "Bool" since `"ast.Exp."` is indeed a prefix of
+-- "ast.Exp.bool". Check types.lua for more usage examples.
+--
+-- @param tag: The type name (string)
+-- @param tag_prefix: The prefix to test (string)
+function typedecl.match_tag(tag, tag_prefix)
+    return type(tag) == "string" and string.match(tag, tag_prefix .. ".(.*)")
+end
```

Comment:
One problem with using string.match here is that special characters in the tag_prefix will be treated as a regular expression. For example, a prefix of `types.T` would also match `typesaT` or `typesbT`, which is not what we want to happen.

One possible fix is to use string.sub to get the parts of the string that we are looking for:

```

function typedecl.match_tag(tag, tag_prefix)
    local n = #tag_prefix
    if type(tag) == "string" and
       string.sub(tag, 1, n) == tag_prefix and
       string.byte(tag, n+1) == 46 -- "."
    then
        return string.sub(tag, n+2)
    else
        return false
    end
end
```